source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,153 @@
+ }
+
+In most cases while implementing an abstract Manipulator you want to have two constructors:
+
+* One without arguments (no-args) which calls the second
+* and one single-argument constructor which is actually used for the provided ``value``
+
+Now, after the Manipulator is done you have to check if you are extending ``AbstractData`` or not.
+Remember that we decided to extend ``AbstractBooleanData`` above.
+If you did too, you must now call ``registerFieldGetter()`` and ``registerFieldSetter`` in your constructor.
+Don't forget the ``Field`` registration!
+
+.. code-block:: java
+
+ private void registerStuff() {
+      registerFieldGetter(Keys.IS_FLYING, new GetterFunction<Object>() {
+             @Override
+             public Value get() {
+                 return getflying();
+             }
+         });
+      registerFieldSetter(Keys.IS_FLYING, new SetterFunction<Object>() {
+              @Override
+              public void set(Object value) {
+                   setFlying(((Boolean) value).booleanValue());
+              }
+          });
+  }
+
+That's it. The ``DataManipulator`` should be done now.
+
+2. ImmutableDataManipulator
+===========================
+
+Just repeat the above mentioned steps for the ``ImmutableDataManipulator``.
+
+Now, after the ``Manipulators`` are done we have to register the ``KEY`` and finally implement the
+``DataProcessor`` and ``ValueProcessor``.
+
+3. Register the Key in the KeyRegistry
+======================================
+
+The next step is to register your ``KEYS`` in the ``KeyRegistry``:
+
+.. code-block:: java
+
+ public static void registerKeys() {
+      keyMap.put("IS_FLYING", makeSingleKey(Boolean.class, Value.class, of("IsFlying")));
+ }
+
+
+4. Implement the DataProcessor
+==============================
+
+Next up is the ``DataProcessor``. A ``DataProcessor`` implements your ``DataManipulator`` to work semi-natively with
+Minecraft's objects. Unfortunately, since it's always easier to just directly implement all of the processing of a
+particular set of data in one place, we have ``DataProcessors`` to do just that.
+Currently, a ``DataProcessor`` has the following methods:
+
+.. code-block:: java
+
+ public interface DataProcessor<M extends DataManipulator<M, I>, I extends ImmutableDataManipulator<I, M>> {
+
+    int getPriority();
+
+    boolean supports(DataHolder dataHolder);
+
+    Optional<M> from(DataHolder dataHolder);
+
+    Optional<M> createFrom(DataHolder dataHolder);
+
+    Optional<M> fill(DataHolder dataHolder, M manipulator, MergeFunction overlap);
+
+    Optional<M> fill(DataContainer container, M m);
+
+    DataTransactionResult set(DataHolder dataHolder, M manipulator, MergeFunction function);
+
+    Optional<I> with(Key<? extends BaseValue<?>> key, Object value, I immutable);
+
+    DataTransactionResult remove(DataHolder dataHolder);
+ }
+
+The first thing that we notice is a ``getPriority()`` method. This is used for multi-processor registration, or in more
```

Comment:
It would be nice to have this in sub sections or like bulletted points of this section or something.